### DISCUSSION TITLE
Design question for control plane HA and Services HA

### DISCUSSION THREAD

**Discussion by philuxe** [kube-vip/kube-vip#1002] [Q&A] [answered]
Hi all,

I'm wondering how I can achieve both Control Plane and Services HA with kubevip knowing I dont want traffic to services flowing via control plan nodes. The doc was not 100% clear to me on this.

I have a 5 nodes cluster (baremetal) with 3 cp and 2 workers nodes. I have installed kubevip as a static pod and was able to get Control Plane HA. Problem is any traffic going to Loadbalancer services goes via tha control plane which I would like to avoid for permformance and control plan protection reasons.

Should I install another kubevip instance as daemon set for this purpose ? how both kubevip instances (static pod + DS) will coexist ?

Or Should I have static pods on worker nodes too ?

**Comment by thebsdbox** [comment, marked as answer]
The best practice is:
- 3x Static pods on control plane nodes
- Daemonset on workers for services

**Reply by LMSSonos** [reply]
@thebsdbox may I ask something in addition. I have my homelab, 4 nodes, all controller & worker. I have a few short questions which I'm not really sure

1) The IP passed in `--tls-san` in k3s - should I use this for the static pod or the daemonset?
2) When I have multiple services like pihole,  etc, where I want a own VIP for the service, do I have to create a daemonset for each VIP? Do I need to configure them somewhere else in addition?
3) https://kube-vip.io/docs/installation/static/#kube-vip-as-ha-load-balancer-or-both mentions that using `--controlplane` and `--service` makes a `highly available Kubernetes cluster` - but you recommend to run static pod with `--controlplane` and then Deamonesets with `--service` - is this correct? 

Thank you very much

**Reply by Cellebyte** [reply]
This is correct.

kube-vip supports an annotation on the Service Object where you can specify the used IP address. Either as SingleStack or even as DualStack. I already mentioned the annotation in your Issue -> #1016

**Reply by nbav12** [reply]
Thank you @thebsdbox for your reply.

I want to clarify this, because the docs are not 100% clear to me.

I use kube-vip only for Services (type: `LoadBalancer`), with the `--services` flag.

I deploy kube-vip as a DaemonSet, using the `--inCluster` and `--taint` flags, because this is what the [documentation ](https://kube-vip.io/docs/installation/flags/) recommends for a DaemonSet. However, the `--taint` flag specifically configures kube-vip to run only on control-plane nodes.

But, according to what you wrote:

> Daemonset on workers for services

Is it possible to run kube-vip on the worker nodes instead? I can't find any reference to this setup in the documentation.

IMHO, running kube-vip (only for Services) on the workers as a DaemonSet makes more sense than running it on the control-plane nodes, since the workloads exposed by the Services reside on the workers. This should also save us from having to hop between nodes, shouldn't it?

If so, can I safely remove the `--taint` flag (or any related configuration) and instead configure a custom taint/toleration setup for the worker nodes?

**Reply by Cellebyte** [reply]
Just remove the taint and it should not add the toleration for control-plane.
